astropy/utils/parsing.py

```diff
@@ -0,0 +1,128 @@
+# -*- coding: utf-8 -*-
+# Licensed under a 3-clause BSD style license - see LICENSE.rst
+"""
+Wrappers for PLY to provide thread safety.
+"""
+
+_TAB_HEADER = """# -*- coding: utf-8 -*-
+# Licensed under a 3-clause BSD style license - see LICENSE.rst
+
+# This file was automatically generated from ply. To re-generate this file,
+# remove it from this folder, then build astropy and run the tests in-place:
+#
+#   python setup.py build_ext --inplace
+#   pytest {package}
+#
+# You can then commit the changes to this file.
+
+"""
+
+import contextlib
+import functools
+import re
+import os
+import threading
+
+
+_LOCK = threading.RLock()
+
+
+def _add_tab_header(filename, package):
+    with open(filename, 'r') as f:
+        contents = f.read()
+
+    with open(filename, 'w') as f:
+        f.write(_TAB_HEADER.format(package=package))
+        f.write(contents)
+
+
+@contextlib.contextmanager
+def _patch_get_caller_module_dict(module):
+    """Temporarily replace the module's get_caller_module_dict.
+
+    It's wrapped to offset the lookup depth by one level, so that we get the
+    function which defines the symbols instead of the lex/yacc wrapper in
+    this module.
+    """
+    original = module.get_caller_module_dict
+
+    @functools.wraps(original)
+    def wrapper(levels):
+        # Add 2, not 1, because the wrapper itself adds another level
+        return original(levels + 2)
+
+    module.get_caller_module_dict = wrapper
+    yield
+    module.get_caller_module_dict = original
+
+
+def lex(lextab, package, reflags=int(re.VERBOSE)):
+    """Create a lexer from local variables.
+
+    It automatically compiles the lexer in optimized mode, writing to
+    `lextab` in the same directory as the calling file.
+
+    This function is thread-safe. The returned lexer is *not* thread-safe, but
+    if it is used exclusively with a single parser returned by :func:`yacc`
+    then it will be safe.
+
+    The `package` is inserted into the instructions in the generated file that
+    explain which package to test to rebuild it.
+
+    It is only intended to work with callbacks defined in a closure, not a class
```

Comment:
I'm slightly confused. Does `closure` refer to the class method scope the parsers are defined in inside the various classes?

Reply:
I've reworded it slightly. http://www.dabeaz.com/ply/ply.html#ply_nn17 might help.

Reply:
Nice!